src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentSplatLoweringPass.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Runtime.InteropServices.ComTypes;
+using Microsoft.AspNetCore.Razor.Language.Intermediate;
+
+namespace Microsoft.AspNetCore.Razor.Language.Components
+{
+    internal class ComponentSplatLoweringPass : ComponentIntermediateNodePassBase, IRazorOptimizationPass
+    {
+        // Run after component lowering pass
+        public override int Order => 50;
+
+        protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            if (!IsComponentDocument(documentNode))
+            {
+                return;
+            }
+
+            var references = documentNode.FindDescendantReferences<TagHelperPropertyIntermediateNode>();
+            for (var i = 0; i < references.Count; i++)
+            {
+                var reference = references[i];
+                var node = (TagHelperPropertyIntermediateNode)reference.Node;
+
+                if (node.TagHelper.IsSplatTagHelper() && node.IsDirectiveAttribute)
+                {
+                    reference.Replace(RewriteUsage(reference.Parent, node));
+                }
+            }
+        }
+
+        private IntermediateNode RewriteUsage(IntermediateNode parent, TagHelperPropertyIntermediateNode node)
+        {
+            if (parent is ComponentIntermediateNode)
+            {
+                var result = new ComponentAttributeIntermediateNode()
+                {
+                    IsAttributeSplat = true,
+                    Source = node.Source,
+                    TypeName = ComponentsApi.AddMultipleAttributesTypeFullName, // Needed for type-inference cases
+                };
+
+                result.Children.AddRange(node.Children);
+                result.Diagnostics.AddRange(node.Diagnostics);
+
+                return result;
+            }
+            else
+            {
+                var result = new HtmlAttributeIntermediateNode()
```

Comment:
I don't get when this would happen.

Reply:
me either but I didn't want to throw. Do you have another suggestion?

Reply:
I'm not going to throw. That's a VS crash bruh.

Reply:
If I could do it all again, I would not make the IR weakly typed. We've had many crashing bugs due to not being able to predict which IR nodes we will see.

Reply:
These days I tend to take the position of being OK with crashes if they enlighten us to new scenarios. Otherwise we have this piece of code that's never hit, never tested and likely to result in bugs that are harder to catch.